File: experiment-14/runs/language=clojure_model=sonnet_prompt=ATDD/rep1/brazilian-soccer-mcp/test/brazilian_soccer_mcp/acceptance_test.clj
(ns brazilian-soccer-mcp.acceptance-test
  "Executable acceptance tests for the Brazilian Soccer MCP server.
   Tests exercise the system through its public MCP tool interface only.
   Each test starts from a clean, freshly-loaded data state."
  (:require [clojure.test :refer [deftest is testing use-fixtures]]
            [brazilian-soccer-mcp.tools :as tools]
            [brazilian-soccer-mcp.data :as data]))

;; ---------------------------------------------------------------------------
;; Test fixture: load data once per namespace
;; ---------------------------------------------------------------------------

(def ^:private test-data-path
  (-> (java.io.File. ".")
      (.getCanonicalPath)
      (str "/../data/kaggle")))

(defonce ^:private loaded-data
  (delay (data/load-all-data test-data-path)))

(defn- get-data []
  @loaded-data)

;; ---------------------------------------------------------------------------
;; 1. Match Queries
;; ---------------------------------------------------------------------------

(deftest find-matches-by-two-teams
  (testing "Can find all matches between Flamengo and Fluminense"
    (let [result (tools/find-matches (get-data)
                                     {:team1 "Flamengo" :team2 "Fluminense"})]
      (is (map? result))
      (is (contains? result :matches))
      (is (seq (:matches result)) "Should find at least one Fla-Flu match")
      (is (every? #(let [home (str (:home-team %))
                         away (str (:away-team %))]
                     (or (and (re-find #"(?i)flamengo" home)
                              (re-find #"(?i)fluminense" away))
                         (and (re-find #"(?i)fluminense" home)
                              (re-find #"(?i)flamengo" away))))
                  (:matches result))
          "All matches should involve both Flamengo and Fluminense")
      (is (every? #(and (contains? % :home-team)
                        (contains? % :away-team)
                        (contains? % :home-goal)
                        (contains? % :away-goal)
                        (contains? % :date)
                        (contains? % :competition))
                  (:matches result))
          "Each match should have required fields"))))

(deftest find-matches-by-single-team
  (testing "Can find all Palmeiras matches in 2023"
    (let [result (tools/find-matches (get-data)
                                     {:team "Palmeiras" :season 2023})]
      (is (map? result))
      (is (seq (:matches result)) "Should find Palmeiras 2023 matches")
      (is (every? #(or (re-find #"(?i)palmeiras" (str (:home-team %)))
                       (re-find #"(?i)palmeiras" (str (:away-team %))))
                  (:matches result))
          "All matches should involve Palmeiras")
      (is (every? #(= (str (:season %)) "2023") (:matches result))
          "All matches should be from 2023"))))

(deftest find-matches-by-competition
  (testing "Can find Copa do Brasil matches"
    (let [result (tools/find-matches (get-data)
                                     {:competition "copa-do-brasil" :season 2023})]
      (is (seq (:matches result)) "Should find Copa do Brasil 2023 matches")
      (is (every? #(re-find #"(?i)copa|brasil|cup" (str (:competition %)))
                  (:matches result))
          "All matches should be from Copa do Brasil")))

  (testing "Can find Brasileirao matches"
    (let [result (tools/find-matches (get-data)
                                     {:competition "brasileirao" :season 2022})]
      (is (seq (:matches result)) "Should find Brasileirao 2022 matches")
      (is (> (count (:matches result)) 100)
          "Should find many Brasileirao matches")))

  (testing "Can find Copa Libertadores matches"
    (let [result (tools/find-matches (get-data)
                                     {:competition "libertadores" :season 2019})]
      (is (seq (:matches result)) "Should find Libertadores 2019 matches"))))

(deftest find-matches-by-date-range
  (testing "Can find matches within a date range"
    (let [result (tools/find-matches (get-data)
                                     {:date-from "2023-01-01" :date-to "2023-06-30"})]
      (is (seq (:matches result)) "Should find matches in first half of 2023")
      (is (every? #(let [d (str (:date %))]
                     (and (>= (compare d "2023-01-01") 0)
                          (<= (compare d "2023-06-30") 0)))
                  (:matches result))
          "All matches should be within the date range"))))

;; ---------------------------------------------------------------------------
;; 2. Team Queries
;; ---------------------------------------------------------------------------

(deftest get-team-home-record
  (testing "Can retrieve Corinthians home record in 2022 Brasileirao"
    (let [result (tools/get-team-stats (get-data)
                                       {:team "Corinthians"
                                        :competition "brasileirao"
                                        :season 2022
                                        :venue "home"})]
      (is (map? result))
      (is (contains? result :team))
      (is (contains? result :wins))
      (is (contains? result :draws))
      (is (contains? result :losses))
      (is (contains? result :goals-for))
      (is (contains? result :goals-against))
      (is (contains? result :matches-played))
      (is (= (+ (:wins result) (:draws result) (:losses result))
             (:matches-played result))
          "W+D+L should equal matches played")
      (is (pos? (:matches-played result))
          "Should have played some home matches"))))

(deftest get-team-overall-stats
  (testing "Can retrieve overall team statistics"
    (let [result (tools/get-team-stats (get-data)
                                       {:team "Flamengo"
                                        :competition "brasileirao"
                                        :season 2019})]
      (is (pos? (:matches-played result)))
      (is (number? (:win-rate result)))
      (is (<= 0.0 (:win-rate result) 1.0)
          "Win rate should be between 0 and 1"))))

;; ---------------------------------------------------------------------------
;; 3. Player Queries
;; ---------------------------------------------------------------------------

(deftest find-players-by-nationality
  (testing "Can find Brazilian players in the dataset"
    (let [result (tools/find-players (get-data)
                                     {:nationality "Brazil" :limit 500})]
      (is (map? result))
      (is (contains? result :players))
      (is (seq (:players result)) "Should find Brazilian players")
      (is (> (count (:players result)) 100)
          "Should find many Brazilian players")
      (is (every? #(contains? % :name) (:players result))
          "Each player should have a name")
      (is (every? #(contains? % :overall) (:players result))
          "Each player should have an overall rating")
      (is (every? #(contains? % :club) (:players result))
          "Each player should have a club"))))

(deftest find-players-by-name
  (testing "Can search players by name"
    (let [result (tools/find-players (get-data)
                                     {:name "Gabriel"})]
      (is (seq (:players result)) "Should find players named Gabriel")
      (is (every? #(re-find #"(?i)gabriel" (str (:name %)))
                  (:players result))
          "All results should have 'Gabriel' in the name"))))

(deftest find-players-by-club
  (testing "Can find players at Santos FC (has entries in FIFA data)"
    (let [result (tools/find-players (get-data)
                                     {:club "Santos"})]
      (is (seq (:players result)) "Should find Santos players")
      (is (every? #(re-find #"(?i)santos" (str (:club %)))
                  (:players result))
          "All results should be Santos players")))

  (testing "Can find players by position and club"
    (let [result (tools/find-players (get-data)
                                     {:club "Santos" :position "ST"})]
      (is (every? #(re-find #"(?i)santos" (str (:club %)))
                  (:players result))
          "All results should be Santos players"))))

(deftest find-top-rated-players
  (testing "Can find top-rated Brazilian players sorted by overall"
    (let [result (tools/find-players (get-data)
                                     {:nationality "Brazil"
                                      :min-overall 85
                                      :sort-key "overall"
                                      :limit 10})]
      (is (<= (count (:players result)) 10) "Should respect limit")
      (is (every? #(>= (:overall %) 85) (:players result))
          "All players should meet min-overall threshold")
      (let [ratings (map :overall (:players result))]
        (is (= ratings (sort > ratings))
            "Players should be sorted by overall rating descending")))))

;; ---------------------------------------------------------------------------
;; 4. Head-to-Head Queries
;; ---------------------------------------------------------------------------

(deftest get-head-to-head-record
  (testing "Can get head-to-head record between two teams"
    (let [result (tools/get-head-to-head (get-data)
                                         {:team1 "Flamengo"
                                          :team2 "Corinthians"})]
      (is (map? result))
      (is (contains? result :team1))
      (is (contains? result :team2))
      (is (contains? result :team1-wins))
      (is (contains? result :team2-wins))
      (is (contains? result :draws))
      (is (contains? result :total-matches))
      (is (= (+ (:team1-wins result) (:team2-wins result) (:draws result))
             (:total-matches result))
          "W1+W2+Draws should equal total matches")
      (is (pos? (:total-matches result))
          "Should find matches between these historic rivals"))))

;; ---------------------------------------------------------------------------
;; 5. Competition Standings
;; ---------------------------------------------------------------------------

(deftest get-brasileirao-standings
  (testing "Can calculate 2019 Brasileirao standings"
    (let [result (tools/get-standings (get-data)
                                      {:season 2019
                                       :competition "brasileirao"})]
      (is (map? result))
      (is (contains? result :standings))
      (is (seq (:standings result)) "Should have standings entries")
      (let [standings (:standings result)]
        (is (every? #(and (contains? % :team)
                          (contains? % :points)
                          (contains? % :wins)
                          (contains? % :draws)
                          (contains? % :losses)
                          (contains? % :goals-for)
                          (contains? % :goals-against)
                          (contains? % :position))
                    standings)
            "Each standing entry should have required fields")
        (let [points (map :points standings)]
          (is (= points (sort > points))
              "Standings should be sorted by points descending"))
        (is (= 1 (:position (first standings)))
            "First team should have position 1")
        (is (re-find #"(?i)flamengo" (str (:team (first standings))))
            "Flamengo won the 2019 Brasileirao")))))

(deftest get-historical-standings
  (testing "Can calculate standings from historical data"
    (let [result (tools/get-standings (get-data)
                                      {:season 2003
                                       :competition "brasileirao"})]
      (is (seq (:standings result))
          "Should compute standings for 2003 season"))))

;; ---------------------------------------------------------------------------
;; 6. Statistical Analysis
;; ---------------------------------------------------------------------------

(deftest get-biggest-wins
  (testing "Can find biggest victories in the dataset"
    (let [result (tools/get-statistics (get-data)
                                       {:stat-type "biggest-wins"
                                        :limit 10})]
      (is (map? result))
      (is (contains? result :results))
      (let [wins (:results result)]
        (is (<= (count wins) 10) "Should respect limit")
        (is (seq wins) "Should find some big wins")
        (let [margins (map #(Math/abs (- (:home-goal %) (:away-goal %))) wins)]
          (is (= margins (sort > margins))
              "Wins should be sorted by margin descending"))))))

(deftest get-average-goals
  (testing "Can calculate average goals per match for Brasileirao"
    (let [result (tools/get-statistics (get-data)
                                       {:stat-type "goals-per-match"
                                        :competition "brasileirao"})]
      (is (map? result))
      (is (contains? result :average-goals-per-match))
      (is (number? (:average-goals-per-match result)))
      (is (< 1.0 (:average-goals-per-match result) 6.0)
          "Average goals should be a realistic value"))))

(deftest get-home-away-record
  (testing "Can calculate home vs away win rates"
    (let [result (tools/get-statistics (get-data)
                                       {:stat-type "home-away-record"
                                        :competition "brasileirao"})]
      (is (map? result))
      (is (contains? result :home-win-rate))
      (is (contains? result :away-win-rate))
      (is (contains? result :draw-rate))
      (let [total (+ (:home-win-rate result)
                     (:away-win-rate result)
                     (:draw-rate result))]
        (is (< (Math/abs (- total 1.0)) 0.001)
            "Win rates should sum to 1.0")))))

(deftest get-top-scoring-teams
  (testing "Can find the top scoring teams in a season"
    (let [result (tools/get-statistics (get-data)
                                       {:stat-type "top-scoring-teams"
                                        :competition "brasileirao"
                                        :season 2023
                                        :limit 5})]
      (is (contains? result :results))
      (let [teams (:results result)]
        (is (<= (count teams) 5) "Should respect limit")
        (is (every? #(and (contains? % :team)
                          (contains? % :goals))
                    teams)
            "Each entry should have team and goals")
        (let [goals (map :goals teams)]
          (is (= goals (sort > goals))
              "Teams should be sorted by goals descending"))))))

;; ---------------------------------------------------------------------------
;; 7. Data coverage — all 6 CSV files load and are queryable
;; ---------------------------------------------------------------------------

(deftest all-data-files-loaded
  (testing "All 6 CSV datasets are loaded and have records"
    (let [d (get-data)]
      (is (seq (:brasileirao-matches d))
          "Brasileirao matches should be loaded")
      (is (seq (:cup-matches d))
          "Copa do Brasil matches should be loaded")
      (is (seq (:libertadores-matches d))
          "Libertadores matches should be loaded")
      (is (seq (:br-football d))
          "BR Football Dataset should be loaded")
      (is (seq (:historical-brasileirao d))
          "Historical Brasileirao (2003-2019) should be loaded")
      (is (seq (:fifa-players d))
          "FIFA player data should be loaded")
      (is (>= (count (:brasileirao-matches d)) 4000)
          "Should have ~4180 Brasileirao matches")
      (is (>= (count (:cup-matches d)) 1300)
          "Should have ~1337 Cup matches")
      (is (>= (count (:libertadores-matches d)) 1200)
          "Should have ~1255 Libertadores matches")
      (is (>= (count (:br-football d)) 10000)
          "Should have ~10296 BR Football records")
      (is (>= (count (:historical-brasileirao d)) 6800)
          "Should have ~6886 historical matches")
      (is (>= (count (:fifa-players d)) 18000)
          "Should have ~18207 FIFA players"))))

(deftest team-name-normalization
  (testing "Team name variations are handled correctly"
    (let [result1 (tools/find-matches (get-data) {:team "Palmeiras"})
          result2 (tools/find-matches (get-data) {:team "Palmeiras-SP"})]
      (is (seq (:matches result1)) "Should find Palmeiras by short name")
      (is (seq (:matches result2)) "Should find Palmeiras by name with state suffix")
      (is (= (count (:matches result1)) (count (:matches result2)))
          "Both name forms should return the same matches"))))

(deftest cross-file-player-and-match-query
  (testing "Can correlate player data with match data (Brazilian players + Flamengo matches)"
    (let [players (tools/find-players (get-data) {:nationality "Brazil" :limit 20})
          matches (tools/find-matches (get-data) {:team "Flamengo" :season 2019})]
      (is (seq (:players players)) "Should find Brazilian players")
      (is (seq (:matches matches)) "Should find Flamengo 2019 matches")
      (is (number? (count (:players players))))
      (is (number? (count (:matches matches)))))))

File: experiment-13/runs/language=go_model=sonnet_prompt=ATDD/rep1/acceptance_test.go
package main

import (
	"context"
	"os"
	"strings"
	"testing"

	"github.com/mark3labs/mcp-go/mcp"
	"brazilian-soccer-mcp/soccer"
	"brazilian-soccer-mcp/tools"
)

var testStore *soccer.Store

func TestMain(m *testing.M) {
	var err error
	testStore, err = soccer.LoadStore("data/kaggle")
	if err != nil {
		panic("failed to load test data: " + err.Error())
	}
	os.Exit(m.Run())
}

func callTool(t *testing.T, handler func(context.Context, mcp.CallToolRequest) (*mcp.CallToolResult, error), args map[string]interface{}) string {
	t.Helper()
	req := mcp.CallToolRequest{}
	req.Params.Arguments = args
	result, err := handler(context.Background(), req)
	if err != nil {
		t.Fatalf("tool error: %v", err)
	}
	if result.IsError {
		t.Fatalf("tool returned error")
	}
	for _, c := range result.Content {
		if tc, ok := c.(mcp.TextContent); ok {
			return tc.Text
		}
	}
	return ""
}

// AT-01: Find matches between two teams
func TestAT01_FindMatchesBetweenTwoTeams(t *testing.T) {
	h := tools.NewFindMatchesHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"team1": "Flamengo",
		"team2": "Fluminense",
	})
	if !strings.Contains(strings.ToLower(result), "flamengo") {
		t.Errorf("expected result to contain Flamengo, got: %s", result)
	}
	if !strings.Contains(strings.ToLower(result), "fluminense") {
		t.Errorf("expected result to contain Fluminense, got: %s", result)
	}
	// Should have multiple matches
	lines := strings.Split(strings.TrimSpace(result), "\n")
	if len(lines) < 3 {
		t.Errorf("expected multiple match results, got %d lines", len(lines))
	}
}

// AT-02: Find matches for a team in a specific season
func TestAT02_FindMatchesForTeamInSeason(t *testing.T) {
	h := tools.NewFindMatchesHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"team":   "Palmeiras",
		"season": float64(2018),
	})
	if !strings.Contains(strings.ToLower(result), "palmeiras") {
		t.Errorf("expected result to contain Palmeiras, got: %s", result)
	}
	if !strings.Contains(result, "2018") {
		t.Errorf("expected result to contain season 2018, got: %s", result)
	}
}

// AT-03: Find matches by competition (Libertadores for a Brazilian team)
func TestAT03_FindMatchesByCompetition(t *testing.T) {
	h := tools.NewFindMatchesHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"team":        "Flamengo",
		"competition": "libertadores",
	})
	if !strings.Contains(strings.ToLower(result), "flamengo") {
		t.Errorf("expected result to contain Flamengo, got: %s", result)
	}
	if !strings.Contains(strings.ToLower(result), "libertadores") && !strings.Contains(strings.ToLower(result), "group stage") && !strings.Contains(strings.ToLower(result), "knockout") {
		// Just check we got matches back
		if len(result) < 10 {
			t.Errorf("expected Libertadores matches for Flamengo, got: %s", result)
		}
	}
}

// AT-04: Get team statistics
func TestAT04_GetTeamStats(t *testing.T) {
	h := tools.NewGetTeamStatsHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"team":        "Corinthians",
		"competition": "brasileirao",
	})
	if !strings.Contains(strings.ToLower(result), "corinthians") {
		t.Errorf("expected result to contain Corinthians, got: %s", result)
	}
	// Should contain wins/draws/losses stats
	if !strings.Contains(strings.ToLower(result), "win") && !strings.Contains(strings.ToLower(result), "draw") && !strings.Contains(strings.ToLower(result), "loss") {
		t.Errorf("expected result to contain win/draw/loss stats, got: %s", result)
	}
}

// AT-05: Find players by nationality
func TestAT05_FindPlayersByNationality(t *testing.T) {
	h := tools.NewFindPlayersHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"nationality": "Brazil",
		"limit":       float64(10),
	})
	if !strings.Contains(result, "Brazil") {
		t.Errorf("expected result to contain Brazil, got: %s", result)
	}
	lines := strings.Split(strings.TrimSpace(result), "\n")
	if len(lines) < 5 {
		t.Errorf("expected at least 5 Brazilian players, got %d lines", len(lines))
	}
}

// AT-06: Find players by club
func TestAT06_FindPlayersByClub(t *testing.T) {
	h := tools.NewFindPlayersHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"club": "Flamengo",
	})
	if !strings.Contains(result, "Flamengo") {
		t.Errorf("expected result to contain Flamengo, got: %s", result)
	}
}

// AT-07: Get league standings for 2019 Brasileirao
func TestAT07_GetLeagueStandings2019(t *testing.T) {
	h := tools.NewGetStandingsHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"season":      float64(2019),
		"competition": "brasileirao",
	})
	if !strings.Contains(strings.ToLower(result), "flamengo") {
		t.Errorf("expected standings to contain Flamengo, got: %s", result)
	}
	// Flamengo won 2019, should be first or near top
	lines := strings.Split(strings.TrimSpace(result), "\n")
	foundFlamengo := false
	for i, line := range lines {
		if strings.Contains(strings.ToLower(line), "flamengo") {
			if i > 5 { // should be in top positions (first few lines after header)
				t.Errorf("Flamengo should be near top of 2019 standings but was at line %d", i)
			}
			foundFlamengo = true
			break
		}
	}
	if !foundFlamengo {
		t.Errorf("Flamengo not found in 2019 standings: %s", result)
	}
}

// AT-08: Get head-to-head stats
func TestAT08_GetHeadToHeadStats(t *testing.T) {
	h := tools.NewGetHeadToHeadHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"team1": "Flamengo",
		"team2": "Fluminense",
	})
	if !strings.Contains(strings.ToLower(result), "flamengo") {
		t.Errorf("expected result to contain Flamengo, got: %s", result)
	}
	if !strings.Contains(strings.ToLower(result), "fluminense") {
		t.Errorf("expected result to contain Fluminense, got: %s", result)
	}
	// Should contain numerical stats
	if !strings.Contains(result, "win") && !strings.Contains(result, "Win") && !strings.Contains(result, "draw") && !strings.Contains(result, "Draw") {
		t.Errorf("expected h2h stats, got: %s", result)
	}
}

// AT-09: Team name normalization
func TestAT09_TeamNameNormalization(t *testing.T) {
	h := tools.NewFindMatchesHandler(testStore)
	result1 := callTool(t, h, map[string]interface{}{
		"team":   "Palmeiras-SP",
		"season": float64(2018),
	})
	result2 := callTool(t, h, map[string]interface{}{
		"team":   "Palmeiras",
		"season": float64(2018),
	})
	// Both should find matches (normalization should make them equivalent)
	if len(result1) == 0 {
		t.Errorf("searching for Palmeiras-SP returned no results")
	}
	if len(result2) == 0 {
		t.Errorf("searching for Palmeiras returned no results")
	}
	// Both should contain Palmeiras results
	if !strings.Contains(strings.ToLower(result1), "palmeiras") {
		t.Errorf("Palmeiras-SP search should return Palmeiras matches: %s", result1)
	}
}

// AT-10: Get biggest wins
func TestAT10_GetBiggestWins(t *testing.T) {
	h := tools.NewGetStatisticsHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"stat_type": "biggest_wins",
		"limit":     float64(5),
	})
	if len(result) < 10 {
		t.Errorf("expected biggest wins results, got: %s", result)
	}
	// Results should be sorted by goal difference (should have high-scoring matches)
	lines := strings.Split(strings.TrimSpace(result), "\n")
	if len(lines) < 3 {
		t.Errorf("expected at least 3 biggest win results, got %d lines", len(lines))
	}
}

// AT-11: Find players by position
func TestAT11_FindPlayersByPosition(t *testing.T) {
	h := tools.NewFindPlayersHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"position": "ST",
		"limit":    float64(10),
	})
	if len(result) < 10 {
		t.Errorf("expected forward/striker results, got: %s", result)
	}
	lines := strings.Split(strings.TrimSpace(result), "\n")
	if len(lines) < 5 {
		t.Errorf("expected at least 5 forwards, got %d lines", len(lines))
	}
}

// AT-12: Get average goals per match statistic
func TestAT12_GetAverageGoalsStatistic(t *testing.T) {
	h := tools.NewGetStatisticsHandler(testStore)
	result := callTool(t, h, map[string]interface{}{
		"stat_type": "goals_average",
	})
	if !strings.Contains(strings.ToLower(result), "average") && !strings.Contains(strings.ToLower(result), "goals") {
		t.Errorf("expected goals average statistic, got: %s", result)
	}
}
